### DISCUSSION TITLE
inference speed and identity switching

### DISCUSSION THREAD

**Discussion by terestrialcreature** [talmolab/sleap#1126] [Help!]
Hi guys, thank your for awesome work.

I am using the develop version of sleap, not the library 1.2.9 version

I had a question, namely I had some flies tracking (4) and I used that bottom up model with 100 labels (out of 9000 frames) with different scenarios labeled. I only have 5 body parts connected with a narrow tree. As you suggested.

I have read this and other similiar issues: https://github.com/talmolab/sleap/discussions/771

My problems are: 

1. How to ensure there is only 4 tracks, not additional different 4 in the middle of the video.
2. There is a lot of identity switching even within these tracks, besides proof reading (which can only fix for the current video not the future videos) what can be done to fix this?
3. What about inference speed, if we care about this, what are the most important parameters to tweak to get inference speed up? Currently I need almost more time for inference (with KF and other tracking turned on obviously) then I need for training.

![Screenshot from 2023-01-18 21-51-35](https://user-images.githubusercontent.com/122974574/213292527-d963ac86-fab3-409b-965d-9a348f139d89.png)
![Screenshot from 2023-01-18 21-54-10](https://user-images.githubusercontent.com/122974574/213292531-1714485d-5983-4c00-95f6-ba714261e973.png)

**Comment by roomrys** [comment]
Hi @terestrialcreature,

1. We don't have the functionality to only assign to the same 4 tracks at the moment. This is because it would be very difficult to tell when identity switches happen if the track lines (in the seek bar) remain unbroken.
2. The main things that help with identity switching are the options under _Pre-tracker data cleaning_. I see you already set the _Cull to Target Instance Count_ which is perfect. If your animals overlap often, then I would set a higher 0 < _IoU Theshold_ <= 1 (so that instances will not be deleted if they overlap too much). At a cost to memory usage (you may need to decrease the _Elapsed Frame Window_ or switch to greedy matching), you can _Save shifted instances_ which should help in the accuracy of matching instances since we save a reference to the shifted instances (via optical flow) instead of just looking at the non-shifted instance points. Kalman filtering did not seem to help much in previous posts, so I would have a go without the KF.
3. Getting rid of the KF will should help inference speeds. If you are running on a GPU that should help as well. (You could try switching to greedy matching which *should* be faster, although we have it self-implemented in python whereas our hungarian matching uses a scipy function.)

Let us know if any of those suggestions help. Biggest thing I would suggest is to run w/o KF.

Thanks,
Liezl

**Reply by terestrialcreature** [reply]
Hi @roomrys, thank you

1. Would you advise that more labeling and better NN would help here?
2. Also are there any particular parameters available from the CLI that are not in the GUI that might adress any of the questions?

**Reply by roomrys** [reply]
Hi @terestrialcreature,

1. Tracking does not use a NN at the moment (we just optimize a cost matrix based on temporal instance overlap). We have some experimental code to do supervised tracking which I can help instruct/collaborate on how to use - get in touch with me at [lmaree@salk.edu](mailto:lmaree@salk.edu) if you are interested. Just a heads up that this experimental approach would require manually adding tracks to each labeled instance.
2. The [CLI options for tracking](https://sleap.ai/guides/cli.html?highlight=command%20line#inference-and-tracking) which might help that are not exposed in the GUI are
- #### For generating predictions
- - `--max_edge_length_ratio` (bottom-up only)
- - `--dist_penalty_weight` (bottom=up only)
- - `--peak_threshold`
- #### For generating tracks
- - `--tracking.min_new_track_points`
- - `--tracking.min_match_point`

Note: The parameters in the GUI are generally the ones we have found most important/useful.

Are you able to identify if the problem is in prediction first then tracking second? If some predictions are not appearing then obviously, we will have a track break. If the predictions look great, then we should start messing with our tracking parameters. 

Thanks,
Liezl

**Comment by terestrialcreature** [comment]
Dear @roomrys  thank you for this. A series of questions, I hope others will benefit from this too:

Regarding the current effort, without supervised tracking, I have some additional question:

1. If I label in one frame only one animal,concretely the animal that had low score in previous inference, but DONT label the other animal in the same frame, since it was good previously, will this still be a valid training sample?

2. What does the low score mean, so If I see frames with values of 0.2 is this bad? Also if I see frames with score values of 3 is this good?

3. Generally, what do you advise performance wise to be achieved before going with tracking: "ALL frames need to have atleast 0.4 scores and when going through them them for most of the individual body parts there should be somewhat correct predictions?"

4. How are nodes for tracking enumerated, is the first node we have defined in our skeleton node 0, second 1 etc?

5. Question about the size of the receptive field (for both top down and bottom up) I have noticed the bigger the receptive field, the better the performance. I guess your heuristic is, receptive field should not be bigger then the size of the animal correct? But what happens when animals are of different sizes across the video?

6. Robust quantile similarity score vs IoU treshold? Whats the difference here, especially whats the exact definition of quantile similarity score?

7. If we correct the the predictions of poses around frames when tracking mixing up happens, will this help with tracking mix up? Is this the remedy to avoid it? 

8. What about tracking mix ups happening even if the pose estimations of the body parts (around the frames) is correct. What else would be the remedy?

Here is a concrete example for this:


![Screenshot from 2023-01-26 18-53-34](https://user-images.githubusercontent.com/122974574/214912072-03ec5cf6-02e1-4e7a-8004-a2800f1de377.png)

![Screenshot from 2023-01-26 18-53-15](https://user-images.githubusercontent.com/122974574/214912109-76793c87-7627-48b9-92a5-091e0789d645.png)

![Screenshot from 2023-01-26 18-54-14](https://user-images.githubusercontent.com/122974574/214912096-b307e028-2cbd-42bb-80c7-19f21dec5189.png)

**Reply by roomrys** [reply]
Super long answer, hopefully this cleared some things up. Thanks for posting on the Discussions for community purposes!

>1. If I label in one frame only one animal,concretely the animal that had low score in previous inference, but DONT label the other animal in the same frame, since it was good previously, will this still be a valid training sample?

No, you will need to label all animals in the frame or none at all. If you only label some animals in the frame then that will be an incorrect training example.

>2. What does the low score mean, so If I see frames with values of 0.2 is this bad? Also if I see frames with score values of 3 is this good?

TLDR: For the supervised ID models: the lower the score, the lower the probability that that instance is assigned to the correct class/track. For the normal models: the lower the score, the lower the confidence that the `Node`s in the instance were predicted correctly. I traced the instance scores for the supervised ID models below since I was surprised to find we measure different metrics than the existing models! That being said, since the scores are not normalized, to accurately use their relative scores for labeling purposes, all predictions should come from either a supervised ID model or all predictions should come from a normal model.

**For the top down supervised model:** The score you see is the average score of all predicted instances in the frame:
https://github.com/talmolab/sleap/blob/b37b34f1b06d6e4e5d6973ad1aae8709c566bda8/sleap/gui/dataviews.py#L493-L501
The score attribute is defined as the mean of all points in `"instance_scores"`:
https://github.com/talmolab/sleap/blob/9474f07821c08eb8e40a31154d66f87e62cb9615/sleap/nn/inference.py#L4204-L4231
The `"instance_scores"` are either the peak class probabilities (or the class probabilities after optimal grouping):
https://github.com/talmolab/sleap/blob/9474f07821c08eb8e40a31154d66f87e62cb9615/sleap/nn/inference.py#L3803-L3825
The peak class probabilities come from the output of our `"ClassVectorHead"` model head.

>3. Generally, what do you advise performance wise to be achieved before going with tracking: "ALL frames need to have atleast 0.4 scores and when going through them them for most of the individual body parts there should be somewhat correct predictions?"

Since the scores are not normalized, it is difficult to advise a threshold based on scores. Usually, I recommend a visual check that the predictions look good (you can use the scores to jump through frames that SLEAP has determined might have poor predictions). Once the predictions look good to the labeler, you can move onto tracking. If adding more suggestions does not seem to help, then it is plausible that the model configuration may need to be adjusted.

>4. How are nodes for tracking enumerated, is the first node we have defined in our skeleton node 0, second 1 etc?

The default node and track names are zero-indexed. I might need a bit more clarification here: 
1. Which part of the GUI/code are you referring to when you mention "node 0" and "node 1"? 
2. What you mean by "how are nodes for tracking enumerated" - more specifically the "enumerated" part.

>5. Question about the size of the receptive field (for both top down and bottom up) I have noticed the bigger the receptive field, the better the performance. I guess your heuristic is, receptive field should not be bigger then the size of the animal correct? But what happens when animals are of different sizes across the video?

TLDR: The receptive field RF should include what we are trying to find in the network. A bigger than needed RF will just use up more memory. If animals are of different sizes across the video, then (if using bottom-up) set the RF to be the maximum size of the animal or (if using top-down) label the frame with the largest observed animal - the largest label is the metric used to calculate crop size for top-down.

The receptive field RF should include what we are trying to find in the network. For bottom-up we need to find all the body parts and then group body parts into animals - for this we need to see the entire animal to know how to handle the grouping. For top-down, we use two different models - one to find the centroid (a single point/body-part - no need to see the entire animal), and another one that takes in a cropped image and finds all the body parts in that cropped image (no need to see the entire animal to do the grouping since it is a centered crop about the animal's centroid).

>6. Robust quantile similarity score vs IoU treshold? Whats the difference here, especially whats the exact definition of quantile similarity score?

The [robust quantile similarity scoring](https://github.com/talmolab/sleap/pull/1062) allows for taking a quantile of the list of similarity scores, which is robust to having outliers. The quantile should be high, 0.95 is the proposed value, but 0.9 would make it even more robust. The [Intersection over Union (IOU) threshold](https://sleap.ai/api/sleap.nn.tracking.html?highlight=iou%20threshold#sleap.nn.tracking.TrackCleaner.iou_threshold) is the threshold to use when removing overlapping instances over target count; if None, then only use score to determine which instances to remove.

>7. If we correct the the predictions of poses around frames when tracking mixing up happens, will this help with tracking mix up? Is this the remedy to avoid it?

If pose estimation is obviously wrong, then that will have an affect on tracking since we calculate similarity scores based on distances-between/intersection-of instances in sequential frames. If you have just a node being predicted incorrectly, then using a similarity that is more 

>8. What about tracking mix ups happening even if the pose estimations of the body parts (around the frames) is correct. What else would be the remedy?

You can try modifying the tracking settings per the previous suggestions. The best remedy might be to help us develop a better method of tracking by either direct code or making light of the problems you are facing. Currently we have formulated tracking as an optimization problem, but perhaps supervised ID is the better method - we just need to test this through more use cases. This seems to be a good spot to say thanks to @getzze who has helped us out a lot recently by contributing to SLEAP tracking, although there are still many aspects yet to be improved upon.

Thanks,
Liezl
